source/base/NexusApp.cc

```diff
@@ -99,7 +99,7 @@ NexusApp::NexusApp(G4String init_macro): G4RunManager(), gen_name_(""),
   if (gen_name_ == "") {
     G4Exception("[NexusApp]", "NexusApp()", FatalException, "A generator must be specified.");
   }
-  pg->SetGenerator(ObjFactory<G4VPrimaryGenerator>::Instance().CreateObject(gen_name_));
+  pg->SetGenerator(std::unique_ptr<G4VPrimaryGenerator>(ObjFactory<G4VPrimaryGenerator>::Instance().CreateObject(gen_name_)));
```

Comment:
No, please don't do that!

Dramatic TL;DR: you are suggesting that `SetGenerator` should **lie** about (not) taking ownership of a resource. Please don't do that.

C++ smart pointers serve two major purposes:

+ automatic resource management;
+ giving clear messages about resource management, *in the source code itself*.

Specifically, any function that accepts a reference or a raw pointer is stating that it is only *borrowing* the resource and will *not* take part in the lifetime management of that resource; any function that accepts a smart pointer (i.e. has a parameter with some smart pointer type) is stating that it *will take ownership* of the resource it is being given (shared ownership in the case of `shared_ptr` and sole ownership in the case of `unique_ptr`).

On the other side of the coin, a caller passing a reference or bare pointer into a function should be confident that the function will only borrow the resource and not participate in the management of its lifetime, so passing the resource in will have no effect on its lifetime management whatsoever; a caller passing a *smart* pointer can clearly see, without having to refer to any documentation or ***dig through the source of the function being called***, that the lifetime of this resource **will** be managed by the callee.

Your suggestion to pass in a raw pointer and convert it into a smart one inside the implementation, completely violates this principle: the callee **will** take ownership of the resource, but in order to understand this, the caller has to dig through the source code of the callee. This is exactly the sort of pain and confusion that smart pointers are supposed to avoid.

IIRC, the initial objection to using a smart pointer here was that the resource is allocated by a factory, and that factory's product is passed to Geant4, which does not expect smart pointers. I haven't read the rest of the code in detail, but I suspect that the factory *should* return a `unique_ptr`, and its `.get()` method should be used to lend the resource to Geant4. This would be impossible if G4 calls the factory itself directly, and we then extract the pointer from the bowels of G4 later, before passing it to `SetGenerator`. If this is not the case, then I strongly recommend that you have the factory return a `unique_ptr`. In which case, you wouldn't be creating a `unique_ptr` just before passing it in to `SetGenerator` but `std::move`ing one into `SetGenerator`. This sends a clear message: the factory used to own the resource, but we've `move`d it out of there and given ownership to `SetGenerator`. That's the way `uniqe_ptr` was intended to be used, and unless G4 makes it *completely impossible*, do it that way.

[One key feature that Rust introduces is that the compiler will not allow you to lie about the ownership of resources in the way you suggested! The downside of this is that Rust forces you to understand what you are doing, and that takes time and effort ... and who has time to understand the code we're writing, right?]

Reply:
> IIRC, the initial objection to using a smart pointer here was that the resource is allocated by a factory, and that factory's product is passed to Geant4, which does not expect smart pointers. I haven't read the rest of the code in detail, but I suspect that the factory should return a unique_ptr, and its .get() method should be used to lend the resource to Geant4. This would be impossible if G4 calls the factory itself directly, and we then extract the pointer from the bowels of G4 later, before passing it to SetGenerator. If this is not the case, then I strongly recommend that you have the factory return a unique_ptr. In which case, you wouldn't be creating a unique_ptr just before passing it in to SetGenerator but std::moveing one into SetGenerator. This sends a clear message: the factory used to own the resource, but we've moved it out of there and given ownership to SetGenerator. That's the way uniqe_ptr was intended to be used, and unless G4 makes it completely impossible, do it that way.

This would be the best solution, I agree.